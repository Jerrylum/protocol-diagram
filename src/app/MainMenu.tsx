import { Button, Divider, ListItemText, Menu, MenuItem, Typography } from "@mui/material";
import { observer } from "mobx-react-lite";
import MenuIcon from "@mui/icons-material/Menu";
import React from "react";
import { onDownload, onDownloadAs, onNew, onOpen, onSave, onSaveAs } from "../core/InputOutput";
import { action } from "mobx";
import { isMacOS } from "../core/Util";

export const HotkeyTypography = observer((props: { hotkey: string | undefined }) => {
  const { hotkey } = props;

  if (hotkey === undefined) return null;

  if (isMacOS(navigator.userAgent) === false)
    return <Typography variant="body2" color="text.secondary" children={hotkey.replaceAll("Mod", "Ctrl")} />;

  const temp = hotkey
    .replaceAll("Mod", "⌘")
    .replaceAll("Option", "⌥")
    .replaceAll("Ctrl", "⌃")
    .replaceAll("Shift", "⇧")
    .replaceAll("CapsLock", "⇪")
    .replaceAll("ArrowLeft", "←")
    .replaceAll("ArrowRight", "→")
    .replaceAll("ArrowUp", "↑")
    .replaceAll("ArrowDown", "↓")
    .replaceAll("Tab", "⇥")
    .replaceAll("Del", "⌫")
    .replaceAll(" ", "␣")
    .replaceAll("Esc", "") // Hide escape key
    .replaceAll("+", "")
    .replaceAll("Add", "+")
    .replaceAll("Equal", "+")
    .replaceAll("Subtract", "-")
    .replaceAll("Minus", "-");

  const elements: React.ReactElement[] = [];
  temp.split("").forEach((char, index) => {
    elements.push(
      <Typography
        key={index}
        variant="body2"
        color="text.secondary"
        sx={{
          width: "1em",
          textAlign: "center",
          fontFamily: '-apple-system,system-ui,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,sans-serif'
        }}
        children={char}
      />
    );
  });

  return <>{elements}</>;
});

export const MainMenu = observer(() => {
  const menuBtnRef = React.useRef<HTMLButtonElement>(null);
  const [isMenuOpen, setIsMenuOpen] = React.useState(false);

  const onMenuButtonClick = (callback: () => void): (() => void) => {
    return action(() => {
      setIsMenuOpen(false);
      callback();
    });
  };

  return (
    <>
      <Button
        ref={menuBtnRef}
        sx={{
          position: "fixed",
          top: "16px",
          left: "16px",
          minWidth: "32px",
          width: "32px",
          height: "32px",
          border: "1px solid #AAA",
          borderRadius: "4px",
          display: "flex",
          justifyContent: "center",
          alignItems: "center",
          userSelect: "none",
          cursor: "pointer",
          color: "#444"
        }}
        onClick={() => setIsMenuOpen(true)}>
        <MenuIcon />
      </Button>
      <Menu
        anchorEl={menuBtnRef.current}
        open={isMenuOpen}
        onClose={() => {
          setIsMenuOpen(false);
        }}
        sx={{ left: "8px" }}
        anchorOrigin={{ vertical: "top", horizontal: "right" }}
        transformOrigin={{ vertical: "top", horizontal: "left" }}>
        <MenuItem dense onClick={onMenuButtonClick(onNew)}>
          <ListItemText sx={{ marginRight: "1rem" }}>New</ListItemText>
          <HotkeyTypography hotkey="Mod+P" />
        </MenuItem>
        <Divider />
        <MenuItem dense onClick={onMenuButtonClick(onOpen)}>
          <ListItemText sx={{ marginRight: "1rem" }}>Open File</ListItemText>
          <HotkeyTypography hotkey="Mod+O" />
        </MenuItem>
        <Divider />
        <MenuItem dense onClick={onMenuButtonClick(onSave)}>
          <ListItemText sx={{ marginRight: "1rem" }}>Save</ListItemText>
          <HotkeyTypography hotkey="Mod+S" />
        </MenuItem>
        <MenuItem dense onClick={onMenuButtonClick(onSaveAs)}>
          <ListItemText sx={{ marginRight: "1rem" }}>Save As</ListItemText>
          <HotkeyTypography hotkey="Shift+Mod+S" />
        </MenuItem>
        <Divider />
        <MenuItem dense onClick={onMenuButtonClick(onDownload)}>
          <ListItemText sx={{ marginRight: "1rem" }}>Download</ListItemText>
          <HotkeyTypography hotkey="Mod+D" />
        </MenuItem>
        <MenuItem dense onClick={onMenuButtonClick(onDownloadAs)}>
          <ListItemText sx={{ marginRight: "1rem" }}>Download As</ListItemText>
          <HotkeyTypography hotkey="Shift+Mod+D" />
        </MenuItem>
      </Menu>
    </>
  );
});

