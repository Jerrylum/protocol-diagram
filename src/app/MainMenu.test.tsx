import { act, fireEvent, getByRole, getRoles, render } from "@testing-library/react";
import { HotkeyTypography, MainMenu } from "./MainMenu";

test("HotkeyTypography", () => {
  const components = (
    <>
      <HotkeyTypography hotkey={undefined} />
      <HotkeyTypography hotkey={"Mod+C"} />
      <HotkeyTypography hotkey={"Mod+Option"} />
      <HotkeyTypography hotkey={"Mod+Option+Shift"} />
      <HotkeyTypography hotkey={"Mod+Option+Shift+Ctrl"} />
      <HotkeyTypography hotkey={"CapsLock"} />
      <HotkeyTypography hotkey={"ArrowLeft"} />
      <HotkeyTypography hotkey={"ArrowRight"} />
      <HotkeyTypography hotkey={"ArrowUp"} />
      <HotkeyTypography hotkey={"ArrowDown"} />
      <HotkeyTypography hotkey={"Tab"} />
      <HotkeyTypography hotkey={"Del"} />
      <HotkeyTypography hotkey={" "} />
      <HotkeyTypography hotkey={"Esc"} />
      <HotkeyTypography hotkey={"+"} />
      <HotkeyTypography hotkey={"Add"} />
      <HotkeyTypography hotkey={"Equal"} />
      <HotkeyTypography hotkey={"Subtract"} />
      <HotkeyTypography hotkey={"Minus"} />
    </>
  );

  const result = render(components);
  expect(result.container).toMatchSnapshot();
});

test("HotkeyTypography in MacOS", () => {
  const currentUserAgent = window.navigator.userAgent;
  Object.defineProperty(window.navigator, "userAgent", { value: "Mac", configurable: true });

  const components = (
    <>
      <HotkeyTypography hotkey={undefined} />
      <HotkeyTypography hotkey={"Mod+C"} />
      <HotkeyTypography hotkey={"Mod+Option"} />
      <HotkeyTypography hotkey={"Mod+Option+Shift"} />
      <HotkeyTypography hotkey={"Mod+Option+Shift+Ctrl"} />
      <HotkeyTypography hotkey={"CapsLock"} />
      <HotkeyTypography hotkey={"ArrowLeft"} />
      <HotkeyTypography hotkey={"ArrowRight"} />
      <HotkeyTypography hotkey={"ArrowUp"} />
      <HotkeyTypography hotkey={"ArrowDown"} />
      <HotkeyTypography hotkey={"Tab"} />
      <HotkeyTypography hotkey={"Del"} />
      <HotkeyTypography hotkey={" "} />
      <HotkeyTypography hotkey={"Esc"} />
      <HotkeyTypography hotkey={"+"} />
      <HotkeyTypography hotkey={"Add"} />
      <HotkeyTypography hotkey={"Equal"} />
      <HotkeyTypography hotkey={"Subtract"} />
      <HotkeyTypography hotkey={"Minus"} />
    </>
  );

  const result = render(components);
  expect(result.container).toMatchSnapshot();
});

test("render MainMenu", () => {
  Object.defineProperty(window.navigator, "userAgent", { value: "Mac", configurable: true });

  const components = (
    <div>
      <MainMenu />
    </div>
  );

  const result = render(components);
  const menuBtn = result.container.querySelector("button")!;

  act(() => {
    fireEvent.click(menuBtn);
  });
  expect(document.body).toMatchSnapshot();

  const muiBackdrop = document.body.querySelector(".MuiBackdrop-root")!;

  act(() => {
    fireEvent.click(muiBackdrop);
  });
  expect(document.body).toMatchSnapshot();

  const menuItem1 = document.body.querySelector("li")!;

  act(() => {
    fireEvent.click(menuItem1);
  });
  expect(document.body).toMatchSnapshot();
});
