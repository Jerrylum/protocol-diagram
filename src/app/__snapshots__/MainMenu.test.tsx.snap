// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotkeyTypography 1`] = `
<div>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Ctrl+C
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Ctrl+Option
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Ctrl+Option+Shift
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Ctrl+Option+Shift+Ctrl
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    CapsLock
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    ArrowLeft
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    ArrowRight
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    ArrowUp
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    ArrowDown
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Tab
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Del
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
     
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Esc
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    +
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Add
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Equal
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Subtract
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
  >
    Minus
  </p>
</div>
`;

exports[`HotkeyTypography in MacOS 1`] = `
<div>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⌘
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    C
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⌘
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⌥
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⌘
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⌥
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⇧
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⌘
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⌥
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⇧
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⌃
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⇪
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ←
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    →
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ↑
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ↓
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⇥
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ⌫
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    ␣
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    +
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    +
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    -
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 css-jsrjbk-MuiTypography-root"
  >
    -
  </p>
</div>
`;
